core/sequencer/server.go

```diff
@@ -131,55 +145,68 @@ func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "HighWatermark(): %v", err)
+		return nil, status.Errorf(codes.Internal, "HighWatermarks(): %v", err)
 	}
-
-	// TODO(#1057): If time since last map revision > max timeout, run batch.
-	// TODO(#1047): If time since oldest log item > max latency has elapsed, run batch.
-	// TODO(#1056): If count items > max_batch, run batch.
-
-	// Count items to be processed.  Unfortunately, this means we will be
-	// reading the items to be processed twice.  Once, here in RunBatch and
-	// again in CreateEpoch.
-	sources := make(map[int64]*spb.MapMetadata_SourceSlice)
-	for sliceID, high := range highs {
-		sources[sliceID] = &spb.MapMetadata_SourceSlice{
-			LowestWatermark:  lastMeta.Sources[sliceID].GetHighestWatermark(),
+	sources := make(SourcesEntry)
+	for logID, high := range highs {
+		sources[logID] = &spb.MapMetadata_SourceSlice{
+			LowestWatermark:  lastMeta.Sources[logID].GetHighestWatermark(),
 			HighestWatermark: high,
 		}
 	}
 
-	msgs, err := s.readMessages(ctx, in.DomainId, sources)
-	if err != nil {
-		return nil, err
-	}
-	if len(msgs) < int(in.MinBatch) {
-		return &empty.Empty{}, nil
+	//
+	// Rate limit the creation of new batches.
+	///
+
+	// TODO(#1057): If time since last map revision > max timeout, run batch.
+	// TODO(#1047): If time since oldest queue item > max latency has elapsed, run batch.
+	// If count items >= min_batch, run batch.
+	if count >= in.MinBatch {
+		return s.CreateEpoch(ctx, &spb.CreateEpochRequest{
+			DomainId:    in.DomainId,
+			Revision:    int64(latestMapRoot.Revision) + 1,
+			MapMetadata: &spb.MapMetadata{Sources: sources},
+		})
 	}
 
-	return s.CreateEpoch(ctx, &spb.CreateEpochRequest{
-		DomainId:    in.DomainId,
-		Revision:    int64(latestMapRoot.Revision) + 1,
-		MapMetadata: &spb.MapMetadata{Sources: sources},
-	})
+	// TODO(#1056): If count items == max_batch, should we define the next batch immediately?
+	return &empty.Empty{}, nil
 }
 
-func (s *Server) readMessages(ctx context.Context, domainID string,
-	sources map[int64]*spb.MapMetadata_SourceSlice) ([]*ktpb.EntryUpdate, error) {
+const readBatchSize = int32(1000)
+
+// readMessages returns the full set of EntryUpdates defined by sources.
+// batchSize limits the number of messages to read from a log at one time.
+func (s *Server) readMessages(ctx context.Context, domainID string, sources SourcesEntry,
+	batchSize int32) ([]*ktpb.EntryUpdate, error) {
 	msgs := make([]*ktpb.EntryUpdate, 0)
 	for logID, source := range sources {
-		batch, err := s.logs.ReadLog(ctx, domainID, logID,
-			source.GetLowestWatermark(), source.GetHighestWatermark())
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "ReadQueue(): %v", err)
-		}
-		for _, m := range batch {
-			msgs = append(msgs, &ktpb.EntryUpdate{
-				Mutation:  m.Mutation,
-				Committed: m.ExtraData,
-			})
+		low := source.GetLowestWatermark()
+		// Loop until less than batchSize items are returned.
```

Comment:
Not sure I get why the stopping condition is `count == batchSize`. What if the first log has `batchSize / 2`, and all the others have `batchSize * 100`?

Reply:
This is the looping condition *per log*. 
If the first log has `batchSize / 2`, that means there are no more batches in log 1 to process, we'll proceed to log 2, where we'll retrieve the first `batchSize` items.  Seeing that `count == batchSize`, we'll proceed to fetch the next `batchSize` items until there are less than `batchSize` items left in log 2 to fetch, at which point we'll move on to log 3 etc.